datalad/support/repodates.py

```diff
@@ -0,0 +1,252 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Utilities for checking repository dates.
+"""
+
+import logging
+import operator
+import re
+import time
+
+from six import string_types
+
+from datalad.support.gitrepo import GitRepo, GitCommandError
+
+lgr = logging.getLogger('datalad.repodates')
+
+
+def branch_blobs(repo, branch):
+    """Get all blobs for `branch`.
+
+    Parameters
+    ----------
+    repo : GitRepo
+    branch : str
+
+    Returns
+    -------
+    A generator object that returns (hexsha, content, file name) for each blob
+    in `branch`.  Note: By design a blob isn't tied to a particular file name;
+    the returned file name matches what is returned by 'git rev-list'.
+    """
+    git = repo.repo.git
+    # Note: This might be nicer with rev-list's --filter and
+    # --filter-print-omitted, but those aren't available until Git v2.16.
+    lines = git.rev_list(branch, objects=True).splitlines()
+    # Trees and blobs have an associated path printed.
+    objects = (ln.split() for ln in lines)
+    blob_trees = (obj for obj in objects if len(obj) == 2)
+
+    # This is inefficient.  It makes a git call for each object, some of which
+    # aren't even blobs.  We could instead use 'git cat-file --batch'.
+    for obj, fname in blob_trees:
+        try:
+            yield obj, git.cat_file("blob", obj), fname
+        except GitCommandError:  # The object was a tree.
+            continue
+
+
+def branch_blobs_in_tree(repo, branch):
+    """Get all blobs for the current tree of `branch`.
+
+    Parameters
+    ----------
+    repo : GitRepo
+    branch : str, optional
+
+    Returns
+    -------
+    A generator object that returns (hexsha, content, file name) for each blob.
+    Note: If there are multiple files in the tree that point to the blob, only
+    the first file name that is reported by 'git ls-tree' is used (i.e., one
+    entry per blob is yielded).
+    """
+    seen_blobs = set()
+    git = repo.repo.git
+    tree_lines = git.ls_tree(branch, z=True, r=True)
+    if tree_lines:
+        for line in tree_lines.strip("\0").split("\0"):
+            _, obj_type, obj, fname = line.split()
+            if obj_type == "blob" and obj not in seen_blobs:
+                yield obj, git.cat_file("blob", obj), fname
+            seen_blobs.add(obj)
+
+
+# In uuid.log, timestamps look like "timestamp=1523283745.683191724s" and occur
+# at the end of the line.  In the *.log and *.log.meta files that are
+# associated with annexed files, the timestamps occur at beginning of the line
+# and don't have the "timestamp=" prefix.
```

Comment:
I guess for now it would be fine but may be for future optimization this should be made more explicit?  ATM it seems to take really long and could be because of regex based search instead of just making it more rigid testing... also I wonder if in regex makes sense to state that number of digits should be no less than some number (instead of just `+`)?  didn't time it but may be it could make it faster?